File: CSAlgos/CSAlgosTest/SortingTest.cs
namespace CSAlgosTest;
using CSAlgos;

[TestClass]
public class SortingTests
{
    private static List<int> items = TestUtils.GenerateList(1000, false);

    public void SortTemplate(Func<IList<int>, IList<int>> func)
    {
        var itemsCopy = new List<int>(items);
        var itemsSorted = new List<int>(itemsCopy);
        itemsSorted.Sort();
        var result = func(itemsCopy);
        Assert.IsTrue(itemsSorted.SequenceEqual(result));
    }

    [TestMethod]
    public void InsertionTest() => SortTemplate(Sorting.Insertion<int>);

    [TestMethod]
    public void SelectionTest() => SortTemplate(Sorting.Selection<int>);

    [TestMethod]
    public void MergeTest() => SortTemplate(Sorting.Merge<int>);
}

File: JAlgos/lib/src/test/java/jalgos/SortingTest.java
package jalgos;

import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

class SortingTest {

    private static List<Integer> items = Util.GenerateList(1000, false);

    public void SortTemplate(SortingFunction<List<Integer>, List<Integer>> func)
    {
        var itemsCopy = new LinkedList<Integer>(items);
        var itemsSorted = new ArrayList<Integer>(itemsCopy);
        itemsSorted.sort(null);
        var result = func.sort(itemsCopy);
        assertTrue(itemsSorted.equals(result));
    }

    @Test void insertionTest() {
        SortTemplate((items) -> Sorting.Insertion(items));
    }

    @Test void selectionTest() {
        SortTemplate((items) -> Sorting.Selection(items));
    }

    @Test void mergeTest() {
        SortTemplate((items) -> Sorting.Merge(items));
    }

    @Test void quickTest() {
        SortTemplate((items) -> Sorting.Quick(items));
    }
}
